sdk/python/feast/infra/utils/aws_utils.py

```diff
@@ -0,0 +1,286 @@
+
+    :param redshift_data_client: Redshift Data API Service client
+    :param cluster_id: Redshift Cluster Identifier
+    :param database: Redshift Database Name
+    :param user: Redshift username
+    :param s3_resource: S3 Resource object
+    :param s3_path: S3 path where the Parquet file is temporarily uploaded
+    :param iam_role: IAM Role for Redshift to assume during the COPY command.
+                     The role must grant permission to read the S3 location.
+    :param table_name: The name of the new Redshift table where we copy the dataframe
+    :param df: The Pandas DataFrame to upload
+    """
+    bucket, key = get_bucket_and_key(s3_path)
+
+    # Convert Pandas DataFrame into PyArrow table and compile the Redshift table schema
+    table = pa.Table.from_pandas(df)
+    column_names, column_types = [], []
+    for field in table.schema:
+        column_names.append(field.name)
+        column_types.append(pa_to_redshift_value_type(str(field.type)))
+    column_query_list = ", ".join(
+        [
+            f"{column_name} {column_type}"
+            for column_name, column_type in zip(column_names, column_types)
+        ]
+    )
+
+    # Create the temporary table with the desired schema
+    create_query = f"CREATE TABLE {table_name}({column_query_list})"
+    execute_redshift_statement(
+        redshift_data_client, cluster_id, database, user, create_query
+    )
+
+    # Write the PyArrow Table on disk in Parquet format and upload it to S3
+    with tempfile.TemporaryDirectory() as temp_dir:
+        file_path = f"{temp_dir}/{uuid.uuid4()}.parquet"
+        pq.write_table(table, file_path)
+        s3_resource.Object(bucket, key).put(Body=open(file_path, "rb"))
+
+    # Copy the Parquet file contents to the Redshift table
+    copy_query = (
+        f"COPY {table_name} FROM '{s3_path}' IAM_ROLE '{iam_role}' FORMAT AS PARQUET"
+    )
+    execute_redshift_statement(
+        redshift_data_client, cluster_id, database, user, copy_query
+    )
+
+    # Clean up S3 temporary data
+    s3_resource.Object(bucket, key).delete()
+
+
+def download_s3_directory(s3_resource, bucket: str, key: str, local_dir: str):
+    """ Download the S3 directory to a local disk """
+    bucket_obj = s3_resource.Bucket(bucket)
+    if key != "" and not key.endswith("/"):
+        key = key + "/"
+    for obj in bucket_obj.objects.filter(Prefix=key):
+        local_file_path = local_dir + "/" + obj.key[len(key) :]
+        local_file_dir = os.path.dirname(local_file_path)
+        os.makedirs(local_file_dir, exist_ok=True)
+        bucket_obj.download_file(obj.key, local_file_path)
+
+
+def delete_s3_directory(s3_resource, bucket: str, key: str):
+    """ Delete S3 directory recursively """
+    bucket_obj = s3_resource.Bucket(bucket)
+    if key != "" and not key.endswith("/"):
+        key = key + "/"
+    for obj in bucket_obj.objects.filter(Prefix=key):
+        obj.delete()
+
+
+def unload_redshift_query_to_s3(
```

Comment:
Nitpick, but this function seems like it's doing two separate things

1. Running a query
2. Exporting to S3

Is there a reason for it to be a single function?  I guess it just reads a bit weird. I'm not used to "unloading" a query.

Reply:
Hmm not sure what you mean by this. Running the query by itself unloads the data to S3. I don't do anything besides running the query on my side. Can you elaborate?

Reply:
I think it may be a naming problem. This is technically "run a query and unload the results to S3", but we are calling it "unload query", which is confusing to me.

Reply:
I disagree with "run a query and unload the results to S3". Running a query and unloading the results to S3 are not 2 separate events. Running the query is what unloads the data to S3. It's like saying "select query". In this case it's "unload query". What would you rename the function to?

Reply:
> Running the query is what unloads the data to S3

Yes, because you are executing both the table creation query and table export with a `;` separator. I would have called it something like `execute_redshift_query_and_unload_to_s3`.